exercises/protein-translation/example.go

```diff
@@ -1,57 +1,71 @@
+// Package protein translates RNA sequences into proteins.
 package protein
 
-func FromCodon(codon string) string {
-	switch codon {
+import "errors"
+
+var (
+	STOP           = errors.New("stop")
+	ErrInvalidBase = errors.New("invalid base")
+)
+
+// IsStopCodon checks whether a codon is a stop codon.
+func IsStopCodon(c string) bool {
+	return c == "UAA" || c == "UAG" || c == "UGA"
+
+}
+
+// FromCodon returns the protein for the given codon.
+// If the codon is a stop codon, it returns STOP.
+// If the codon is unrecognized, it returns invalidBase.
+func FromCodon(c string) (string, error) {
+	if IsStopCodon(c) {
+		return "", STOP
+	}
+
+	switch c {
 	case
 		"AUG":
-		return "Methionine"
+		return "Methionine", nil
 	case
 		"UUU",
 		"UUC":
-		return "Phenylalanine"
+		return "Phenylalanine", nil
 	case
 		"UUA",
 		"UUG":
-		return "Leucine"
+		return "Leucine", nil
 	case
 		"UCU",
 		"UCC",
 		"UCA",
 		"UCG":
-		return "Serine"
+		return "Serine", nil
 	case
 		"UAU",
 		"UAC":
-		return "Tyrosine"
+		return "Tyrosine", nil
 	case
 		"UGU",
 		"UGC":
-		return "Cysteine"
+		return "Cysteine", nil
 	case
 		"UGG":
-		return "Tryptophan"
-	case
-		"UAA",
-		"UAG",
-		"UGA":
-		return "STOP"
+		return "Tryptophan", nil
+	default:
+		return "", ErrInvalidBase
 	}
-	return "STOP"
 }
 
+// FromRNA returns the protein sequence for an RNA strand.
 func FromRNA(s string) []string {
-	var res = ""
 	var proteins []string
-	for index, codon := range s {
-		res += string(codon)
-		if index > 0 && (index+1)%3 == 0 {
-			tempCodon := FromCodon(res)
-			if tempCodon == "STOP" {
-				break
-			}
-			proteins = append(proteins, tempCodon)
-			res = ""
+	bases := []rune(s)
+	for i := 0; i < len(bases); i += 3 {
+		p, err := FromCodon(string(bases[i : i+3]))
+		if err == STOP {
```

Comment:
Only nit I have is that we don't ever handle the `ErrInvalidBase` case here. How would we handle a string like the following from the `rnaCase` test slice?

```
OLD
{"UGGUGUUAUUAAUGGUUU", []string{"Tryptophan", "Cysteine", "Tyrosine"}}

NEW
{"UGGUGUUAUUAAUGGABCUUU"}

        protein_translation_test.go:124: FAIL: RNA Translation test: UGGUGUUAUUAAUGGABCUUU
                Expected: ["Tryptophan" "Cysteine"]
                Actual ["Tryptophan" "Cysteine" "Tyrosine"]
```